exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp

```diff
@@ -0,0 +1,229 @@
+    {
+        ThrowNamed("This ILQRSolver can't solve problem of type '" << pointer->type() << "'!");
+    }
+    MotionSolver::SpecifyProblem(pointer);
+    prob_ = std::static_pointer_cast<DynamicTimeIndexedShootingProblem>(pointer);
+
+    // initialize Gain matrices
+    Eigen::MatrixXd x = prob_->get_X(0), u = prob_->get_U(0);
+
+    int T = prob_->get_T();
+    K_gains_.assign(T, Eigen::MatrixXd(u.rows(), x.rows()));
+    Ku_gains_.assign(T, Eigen::MatrixXd(u.rows(), u.rows()));
+    Kv_gains_.assign(T, Eigen::MatrixXd(u.rows(), x.rows()));
+    vk_gains_.assign(T, Eigen::MatrixXd(x.rows(), x.cols()));
+
+    if (debug_) HIGHLIGHT_NAMED("ILQRSolver", "initialized");
+}
+
+void ILQRSolver::BackwardPass()
+{
+    constexpr double min_clamp_ = -1e10;
+    constexpr double max_clamp_ = 1e10;
+    int T = prob_->get_T();
+
+    Eigen::MatrixXd Qf = prob_->get_Qf(), R = prob_->get_R();
+    Eigen::MatrixXd X_star = prob_->get_X_star();
+    DynamicsSolverPtr dynamics_solver = prob_->GetScene()->GetDynamicsSolver();
+
+    // eq. 18
+    vk_gains_[T - 1] = Qf * dynamics_solver->StateDelta(prob_->get_X(T - 1), X_star.col(T - 1));
+    vk_gains_[T - 1] = vk_gains_[T - 1].unaryExpr([min_clamp_, max_clamp_](double x) -> double {
+        return std::min(std::max(x, min_clamp_), max_clamp_);
+    });
+
+    Eigen::MatrixXd Sk = Qf;
+    Sk = Sk.unaryExpr([min_clamp_, max_clamp_](double x) -> double {
+        return std::min(std::max(x, min_clamp_), max_clamp_);
+    });
+
+    for (int i = T - 2; i > 0; i--)
+    {
+        Eigen::MatrixXd x = prob_->get_X(i), u = prob_->get_U(i), Q = prob_->get_Q(i);
+        Eigen::MatrixXd Ak = dynamics_solver->fx(x, u), Bk = dynamics_solver->fu(x, u);
+
+        // eq. 13-16
+        Ak = Ak * dynamics_solver->get_dt() + Eigen::MatrixXd::Identity(Ak.rows(), Ak.cols());
+        Bk = Bk * dynamics_solver->get_dt();
+        // this inverse is common for all factors
+        Eigen::MatrixXd _inv = (Eigen::MatrixXd::Identity(R.rows(), R.cols()) * 1e-5 + R + Bk.transpose() * Sk * Bk).inverse();
+
+        Kv_gains_[i] = _inv * Bk.transpose();
+        K_gains_[i] = _inv * Bk.transpose() * Sk * Ak;
+        Ku_gains_[i] = _inv * R;
+        Sk = Ak.transpose() * Sk * (Ak - Bk * K_gains_[i]) + Q;
+
+        vk_gains_[i] = ((Ak - Bk * K_gains_[i]).transpose() * vk_gains_[i + 1]) -
+                     (K_gains_[i].transpose() * R * u) + (Q * x);
+
+        // fix for large values
+        Sk = Sk.unaryExpr([min_clamp_, max_clamp_](double x) -> double {
+            return std::min(std::max(x, min_clamp_), max_clamp_);
+        });
+        vk_gains_[i] = vk_gains_[i].unaryExpr([min_clamp_, max_clamp_](double x) -> double {
+            return std::min(std::max(x, min_clamp_), max_clamp_);
+        });
+    }
+}
+
+double ILQRSolver::ForwardPass(double alpha, Eigen::MatrixXdRef ref_trajectory)
+{
+    double cost = 0;
+    int T = prob_->get_T();
+    Eigen::VectorXd control_limits = prob_->GetControlLimits();
```

Comment:
`const Eigen::VectorXd&`

Reply:
... and move to dynamics solver. Could also be a matrix with lower and upper bound (for most robots the same though).

Not sure the reference helps since control limits returns it as a copy?